backend/sirius-web-diagrams/src/main/java/org/eclipse/sirius/web/diagrams/components/NodeComponent.java

```diff
@@ -47,72 +51,104 @@ public NodeComponent(NodeComponentProps props) {
-
-            INodeStyle style = nodeDescription.getStyleProvider().apply(nodeVariableManager);
-
-            // @formatter:off
-            var borderNodes = nodeDescription.getBorderNodeDescriptions().stream()
-                    .map(borderNodeDescription -> {
-                        var nodeComponentProps = new NodeComponentProps(nodeVariableManager, borderNodeDescription, true, cache);
-                        return new Element(NodeComponent.class, nodeComponentProps);
-                    })
-                    .collect(Collectors.toList());
-
-            var childNodes = nodeDescription.getChildNodeDescriptions().stream()
-                    .map(childNodeDescription -> {
-                        var nodeComponentProps = new NodeComponentProps(nodeVariableManager, childNodeDescription, false, cache);
-                        return new Element(NodeComponent.class, nodeComponentProps);
-                    })
-                    .collect(Collectors.toList());
-            // @formatter:on
-
-            List<Element> nodeChildren = new ArrayList<>();
-            nodeChildren.add(labelElement);
-            nodeChildren.addAll(borderNodes);
-            nodeChildren.addAll(childNodes);
-
+        if (nodeDescription.isSynchronised()) {
             // @formatter:off
-            NodeElementProps nodeElementProps = NodeElementProps.newNodeElementProps(nodeId)
-                    .type(type)
-                    .targetObjectId(targetObjectId)
-                    .targetObjectKind(targetObjectKind)
-                    .targetObjectLabel(targetObjectLabel)
-                    .descriptionId(nodeDescription.getId())
-                    .borderNode(isBorderNode)
-                    .style(style)
-                    .position(Position.UNDEFINED)
-                    .size(Size.UNDEFINED)
-                    .children(nodeChildren)
-                    .build();
+            nodeDescription.getSemanticElementsProvider().apply(variableManager)
+                .forEach(semanticElement -> children.add(this.renderNodeElementProps(semanticElement, Optional.empty())));
             // @formatter:on
-            Element nodeElement = new Element(NodeElementProps.TYPE, nodeElementProps);
-            children.add(nodeElement);
-
-            cache.put(nodeDescription.getId(), nodeElement);
-            cache.put(semanticElement, nodeElement);
+        } else {
+            Optional<Map> optionalTargetSemanticObjects = variableManager.get(DiagramDescription.TARGET_SEMANTIC_OBJECTS, Map.class);
+            if (optionalTargetSemanticObjects.isPresent()) {
+                Map<String, Object> targetSemanticObjects = optionalTargetSemanticObjects.get();
+               // @formatter:off
+               this.props.getPreviousNodes()
+                   .forEach(prevNode -> {
+                       String targetObjectId = prevNode.getTargetObjectId();
+                       if (targetSemanticObjects.containsKey(targetObjectId)) {
```

Comment:
You are making sure that the targetId from a previous node is contained in TARGET_SEMANTIC_OBJECTS which has been created just a couple methods calls ago using the same previous node. I fail to grasp the meaning of what you are doing and why you are computing this information this way.

Starting with an ADR to properly define the problem and the solution may be helpful.